### PR TITLE
Add snapshot-aware checkpoints and safe WAL slot reuse

### PR DESCRIPTION
**Pull request by JKamsker**
Long-running readers currently prevent WAL checkpointing, and shared-mode readers hold the database mutex throughout enumeration. This change lets checkpoints progress, shared writers commit, and eligible obsolete WAL slots be reused while readers retain their original snapshots—including pages they have not visited.

### Implementation

- Register every collection snapshot atomically with version capture and checkpoint boundary selection. Shared readers use private read-only engines and OS-held lease files, releasing the named writer mutex before enumeration. Crashed readers lose their leases without heartbeat expiry.
- Backfill through the oldest local/shared snapshot. For each page retain the newest version at or below that watermark (its floor), every newer version, and any confirmation records those versions require. A frame being below the watermark alone is insufficient to reclaim it.
- Durably flush data before clearing obsolete frames; durably flush the zeroed slots before publishing reusable capacity. Keep all retained offsets unchanged, invalidate obsolete cache entries, and serialize slot allocation with WAL writes.
- Use appended confirmation positions as stable version IDs, so removing old commits cannot renumber another process's snapshot. Recovery reconstructs holes and required versions without a new file format or persistent free-list file.
- Preserve increasing physical WAL positions per page, including when choosing reusable slots, because older engines checkpoint in physical order. Keep writers, explicit shared transactions, `FOR UPDATE`, and `SELECT INTO` serialized. Isolate sort storage and reject rebuild while readers remain.

### Storage effect and limits

In the process-test workload (two collections of 64 rows with 3,000-byte payloads, 20 updates building history, an open reader, then five updates to the other collection), WAL growth fell from **1,320 KiB to 72 KiB: 94.5% less incremental growth**, for both plain and encrypted files.

This reuses allocated WAL capacity; it does not shrink the live file or punch filesystem holes. Required floors/confirmations remain pinned. Legacy physical replay order also means some free slots cannot serve a particular page, so hot-page writes may still append. Full truncation requires readers to drain. The workload measurement is not a general storage-reduction guarantee.

All concurrent participants must use the shared coordination protocol on the same host with working file-sharing locks. The safety argument, recovery ordering, and constraints are documented in `docs/mvcc-checkpoint.md`.

### Coverage and validation

66 new MVCC cases cover untouched pages, version-zero readers, collection watermarks, registration races, page reuse, concurrent readers/writers/checkpoints, cleanup failures, nested transactions, encrypted sorting, free-slot recovery, and readers paused between offset resolution and page access. Actual child processes exercise different snapshots, serialized writers, killed readers/uncommitted writers, and crashes during backfill, slot clearing, durable flush, free-slot publication, and full WAL reset.

- Release solution build with `TestingEnabled=true` — passed, including .NET Framework compilation.
- Full Release tests with `tests.runsettings`: .NET 8 **1,979 passed / 7 existing skips**; .NET 10 **1,979 passed / 7 existing skips**. All 66 MVCC cases passed on both targets.
- Production library build with `TestingEnabled=false` — passed for netstandard2.0, net8.0, and net10.0.
- `python3 scripts/test-vector-compatibility.py` — passed against LiteDB 5.0.21, plain and encrypted. It now also verifies that 5.0.21 can replay, update, checkpoint, and reopen a reclaimed/reused ordinary v8 WAL, followed by reopening with the current engine.
- Changed-C# size check and `git diff --check` — passed.

Runtime validation was performed on Linux. Windows/macOS execution is left to repository CI. Crash tests cover process termination and write boundaries, not arbitrary torn-sector corruption.

### CODE REVIEW

**Comment by JKamsker** [comment]
## Review findings (head `4a81c61d2`)

**Summary:** I found no correctness bug in the checkpoint/WAL code, but I would not merge yet. It needs a re-merge with dev, a decision on a shared-mode read slowdown, one Windows edge case fixed, and a human review.

### What was checked

- **CI:** all 48 checks pass on the head commit (Windows x86/x64, net462/net481, Linux, ARM64, macOS, cross-process jobs).
- **Merged with current dev:** the branch is 3 commits behind dev, including the shared query IR change (#2944, 239 files), so the green CI does not cover the tree that would actually be merged. Merged locally; the only conflict is `AGENTS.md`. Full suite on Windows, net10, non-UTC time zone: **3066 passed, 0 failed, 7 existing skips**. net8 and net462/net481 were not run on the merged tree.
- **Code read** of the WAL/checkpoint changes: floor retention per page, flush ordering (data flush, then zero the slots, then publish them), confirmation pinning until the whole transaction is obsolete, per-page increasing WAL positions for 5.0.21 replay, lock order (transaction exclusion, then index lock, then writer stream), and lease registration under the mutex all hold up.

### Open items

**1. Shared-mode reads are about 2.1x slower.** Every non-transactional `SharedEngine.Query` now opens a second read-only `LiteEngine` and creates a lease file. Measured against dev (`ffc3edfc9`) vs. this PR merged with dev, Release, net10, Windows, 4 alternating runs, stable:

| Workload (Connection=shared) | dev | PR merged with dev |
|---|---|---|
| 1000 `FindById` | ~1.0 s | ~2.1 s |
| 300 `FindById` + `Update` | ~1.7 s | ~2.05 s |

The PR description does not mention this.

**2. The storage saving is narrower than the description reads.** Direct mode, a reader opened on another collection, then 4000 single-document updates: the WAL ends at 32,000 KiB on both dev and this PR, because every version newer than the oldest reader is retained. The 94.5% figure applies only to history written before the reader opened. Commits in that run were about 6-10% slower than without a reader (the auto-checkpoint body now runs on every commit above the checkpoint threshold, scanning the index under the index write lock).

**3. `Checkpoint()` no longer blocks until the WAL is drained.** It waits 10 ms for exclusivity and otherwise does a partial backfill. A caller that runs `Checkpoint()` and then copies only the data file can get an incomplete backup when a transaction is open, with no signal.

**4. Lease probe on Windows.** `SharedReaderRegistry.OldestVersion()` catches only `IOException`. A delete-pending lease file or an antivirus lock raises `UnauthorizedAccessException`, which escapes into the checkpoint and commit path. Not reproduced; found by reading.

**5. Smaller items**
- The `<db>-readers` directory is never removed.
- The index write lock is held across the checkpoint's fsyncs, which stalls readers' page lookups during a non-exclusive checkpoint.
- A leaked (undisposed) shared reader now grows the WAL silently; before, it blocked on the mutex, which was at least visible.

**6. Process.** Still a draft with no human review; CodeRabbit skipped it for that reason. Crash tests cover process kills and write boundaries, not torn sectors (as the description says).

### Suggested before merge

Re-merge dev and let CI run, decide on item 1 (accept and document, or avoid the second engine for short queries), fix item 4, document items 2 and 3 in the description and release notes, and get a second reviewer on the durability code.

File: LiteDB.Tests/Internals/MvccCheckpoint_Tests.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                catch (Exception ex) { checkpointError = ex; }
+                finally { checkpointDone.Set(); }
+            });
+            reader.Start();
+            try
+            {
+                captured.Wait(TimeSpan.FromSeconds(10)).Should().BeTrue();
+                checkpoint.Start();
+                checkpointStarted.Wait(TimeSpan.FromSeconds(10)).Should().BeTrue();
+                checkpointDone.Wait(TimeSpan.FromMilliseconds(100)).Should().BeFalse();
```
[the remark this answers is not in the document]
Comment (JKamsker) [reply LiteDB.Tests/Internals/MvccCheckpoint_Tests.cs]: Fixed in ee7928743. `TryCheckpoint` now raises a test-only `before-index-lock` stage immediately before `_indexLock.EnterWriteLock()`, and the test sets `checkpointStarted` from that stage, so the 100 ms assertion only starts once the checkpointer is at the lock boundary.

File: LiteDB/Engine/Services/WalIndexService.Checkpoint.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            try
+            {
+                if (--_snapshots[version] == 0) _snapshots.Remove(version);
+            }
+            finally { _indexLock.ExitWriteLock(); }
+        }
+
+        public int Checkpoint() => this.TryCheckpoint(rationed: false);
+
+        public int TryCheckpoint() => this.TryCheckpoint(rationed: false);
```
[the remark this answers is not in the document]
Comment (JKamsker) [reply LiteDB/Engine/Services/WalIndexService.Checkpoint.cs:57]: Not changing this; the partial path is the purpose of the PR, not an oversight.

- `Checkpoint()` making progress while readers are open is what lets the WAL be backfilled and its slots reused when there is always a reader somewhere, which is the production case this PR targets. With `EnterExclusive()` it would again wait for every reader, and with a reader open on the calling thread it would run into the lock timeout. The PR's own tests call `Checkpoint()` with open readers on purpose (`MvccCheckpoint_Tests`, `MvccReclamation_Tests`, `MvccRecovery_Tests`).
- It is not silent any more: `docs/mvcc-checkpoint.md` now states that `Checkpoint()` does not guarantee an empty WAL and what that means for a backup that copies only the data file.
- The legitimate need behind the comment, a caller that must know the WAL drained, is tracked in #2957 as a separate explicit API rather than a change to `Checkpoint()`.

**Comment by JKamsker** [comment]
## Status of the review findings

Follow-up to my review comment above, as of `ee7928743`. The PR description is updated to match.

| Finding | Status |
|---|---|
| Branch behind dev | Merged dev in `9c2a095d3`. Full suite on the merged tree, Windows: net10, net8 and net481 all 0 failed. |
| 1. Shared-mode reads ~2.1x slower | Fixed in `71ee8135c`. Small results complete under the mutex; 1000 `FindById` are back at dev speed (~0.93 s). |
| 2. Saving narrower than described | Partly fixed in `71ee8135c`: versions no live or future snapshot can resolve are now reclaimed, and partial auto-checkpoint work runs on the back-off. Single-page commits under a long-lived reader still append one page each; documented and tracked in #2957. With readers that come and go the WAL stays at about 8 MB. |
| 3. `Checkpoint()` no longer drains the WAL | By design. Documented in `docs/mvcc-checkpoint.md` and the description; explicit API tracked in #2957. |
| 4. Lease probe could throw on Windows | Fixed in `71ee8135c`; an unprobeable lease counts as live. |
| 5. `-readers` directory, index lock across fsyncs, leaked readers | Directory removal fixed in `71ee8135c`. The other two are tracked in #2957. |

CodeRabbit review of `4a81c61d2`:
- Test synchronization at the lock boundary: fixed in `ee7928743`, thread resolved.
- Blocking `Checkpoint()`: declined with reasoning in the thread, resolved.
- Nitpick, lease scan under the index write lock: fixed in `ee7928743`; the scan now runs before the lock.
- Nitpick, empty MVCC probe glob: fixed in `ee7928743`; the test build now fails with a clear message.

**Comment by JKamsker** [comment]
I did a deep review of `ee7928743be87de94a7405b9ce736dcf0868b35c`.

**Verdict: I would block merging on 3 findings.** The central “retain the floor of every live snapshot + newest version + required confirmations” rule looks sound under its coordination assumptions. The problems I found are around preserving those assumptions across recovery, failure cleanup and shared-reader discovery.

## 1. P1 — LiteDB 5.0.21 can resurrect an abandoned reused-slot write

This is the most serious compatibility issue I found.

The PR preserves physical WAL ordering **per page** so 5.0.21 can replay a reclaimed/reused WAL. However, 5.0.21 also relies on an implicit transaction-ID ordering property that slot reuse can violate.

### Failure sequence

Suppose the WAL contains:

```
physical order:

old cold page       tx 1    confirmed
reused hole         tx 41   UNCONFIRMED   cold = 99
existing WAL tail   tx 40   confirmed
```

This is possible if tx 41 writes entirely into reclaimed holes and then crashes/rolls back before writing its confirmation. Per-page physical ordering can still be valid.

Current code correctly restores `_lastTransactionID` as the **maximum observed ID**, including abandoned transactions.

But LiteDB 5.0.21's recovery assigns `_lastTransactionID = transactionID` while scanning physical WAL order. It therefore finishes the above WAL believing the last transaction ID is 40.

The next legacy transaction can reuse **41**. If that unrelated transaction commits, 5.0.21 adds tx 41 to its confirmed transaction set. During checkpoint it then treats **all WAL frames carrying tx 41 as committed**, including the abandoned `cold = 99` frame.

That makes an uncommitted write durable.

This does **not** require old/new engines to run concurrently. It can happen after closing the new version and reopening the file with 5.0.21.

### Why current compatibility coverage misses it

The compatibility fixture tests a **successfully committed** reused-slot transaction before opening with the legacy engine. It does not create the dangerous state where a newer transaction ID exists only in reclaimed holes and has no confirmation at the WAL tail.

### Test I think we need

1. Build reclaimable holes.
2. Start a transaction whose frames all fit into reclaimed slots.
3. Kill/abort it before confirmation.
4. Close all current engines.
5. Open with 5.0.21.
6. Commit an unrelated transaction.
7. Checkpoint and reopen.
8. Assert every abandoned change stayed invisible.
9. Run plain + encrypted.

I would not claim full 5.0.21 write compatibility until this is resolved. Per-page ordering alone is not sufficient.

---

## 2. P1 — Confirmed-flush failure can deadlock against partial reclamation

There is an opposing lock order.

### Checkpoint path

```
WalIndexService.TryCheckpoint
    owns _indexLock (write)
        ...
        ReclaimLogPages
            waits for WAL writer stream lock
```

The PR intentionally holds the index write lock across checkpoint I/O/reclamation.

### Writer failure path

```
DiskService.WriteLogDisk
    owns WAL writer stream lock
        FlushConfirmedLog throws
        _state.Handle(failure)
            engine close / transaction cleanup
                snapshot.ReleaseSnapshotPin()
                    waits for _indexLock (write)
```

So this schedule can deadlock:

```
checkpointer: owns index lock -> waits for WAL writer lock
writer:       owns WAL writer lock -> flush fails -> cleanup waits for index lock
```

Partial checkpointing makes this overlap possible because checkpoint can now do work while a transaction is still alive.

### Fix direction

I think synchronous engine teardown must not happen while the WAL writer monitor is held.

The engine still needs to become immediately unusable after an uncertain confirmed flush, but cleanup/unpinning should happen only after leaving the writer lock.

### Test I think we need

A deterministic barrier test:

- create reclaimable history;
- start checkpoint and stop it after it owns the index lock but before it obtains the writer lock;
- make a concurrent confirmed WAL flush fail while owning the writer lock;
- assert both operations terminate within a bound and recovery is correct.

The existing concurrency and crash-boundary tests do not appear to cover this combined lock/failure schedule.

---

## 3. P1 — Failure to inspect the reader directory can be mistaken for “no readers”

`SharedReaderRegistry.LiveVersions()` starts with:

```csharp
if (!Directory.Exists(_directory)) return _none;
```

Unfortunately `Directory.Exists` returns `false` both when the directory genuinely does not exist **and when determining its existence fails** (for example permission/access errors).

That means:

1. a live shared reader has a lease;
2. writer/checkpointer can still access DB/WAL;
3. probing the lease directory fails;
4. `LiveVersions()` reports zero readers;
5. checkpoint can take the full-reclaim/truncate path;
6. the live reader's fixed WAL offsets are no longer protected.

The per-file handling is conservative, but the directory-existence probe is not.

### Important detail

Returning a synthetic version `0` is **not** a safe generic “unknown reader” marker either.

For page versions 6, 9, 12 where the inaccessible reader actually needs 6, feeding only snapshot 0 into the current reclamation rule retains the newest version 12 but not necessarily 6.

So “unknown registry state” needs to mean **do not reclaim / do not truncate**, not “pretend there is a version-zero reader”.

### Fix direction

Use an inspection path that distinguishes:

- definitely absent registry;
- successfully enumerated registry;
- unknown/unreadable registry.

On unknown: fail the checkpoint or skip reclamation.

Add an injected registry-probe failure test with an actually live external reader.

---

## 4. P2 — Shared query fallback re-executes user callbacks

For ordinary shared queries, `TryBufferResult` executes the query under the mutex. If it exceeds 100 values / 64 KiB, those results are discarded and the query is executed again through the private snapshot engine.

That means public callbacks such as `ReadTransform` run twice for the prefix that was already produced.

For a stateful transform this can create:

- duplicate side effects;
- different values depending on result size;
- exceptions only when a result crosses the buffer threshold.

It can also duplicate much more work than “101 rows” suggests: producing those first rows may already require an expensive scan/sort.

I would at least add a regression test where `ReadTransform` records IDs and detects duplicate execution. Personally I'd skip speculative buffering when a user transform is configured, or find a way to preserve the already-running cursor instead of restarting it.

---

## Additional release concerns

These are mostly already tracked in #2957, but I think they matter before treating this as fully settled:

- **Checkpoint semantics changed:** `Checkpoint()` no longer implies an empty WAL. Data-file-only backup callers need a first-class way to require/verify a drained WAL.
- **Long-lived reader + one-page commits still grows WAL indefinitely** because confirmations append and define versions.
- **Index write lock spans checkpoint I/O**, so partial checkpoint latency can block new snapshots/page lookups.
- **Torn sectors/pages are not covered** by the current crash tests.
- Mixed old/new/direct participants are explicitly unsupported while shared-reader leases exist.

For backups I also would not document “copy both files” as sufficient by itself unless the copy is taken under a consistency protocol preventing incompatible intervening writes/reclamation.

---

## What I think is solid

I did **not** find a counterexample to the core MVCC reclamation rule itself under the documented coordination model:

- snapshot capture + pin publication are serialized with index mutation;
- every live snapshot retains its per-page floor;
- newest versions are retained for future snapshots;
- required confirmation records are retained;
- retained offsets do not move;
- data is flushed before obsolete WAL frames are cleared;
- cleared slots are flushed before becoming reusable.

The paused-reader-after-offset-resolution test is especially valuable coverage.

So I would **not** reject the approach. I would fix the three P1 issues above, add the targeted tests, and then re-review the resulting recovery/locking changes.

**Comment by JKamsker** [comment]
Addressed every finding from the deep review in `bd3ece3e2`:

1. New transaction IDs now get a physical WAL-tail anchor before any later frame can reuse a hole. The compatibility harness leaves an unconfirmed transaction in reclaimed slots, opens with LiteDB 5.0.21, commits an unrelated transaction, checkpoints and reopens; plain and encrypted both pass.
2. A failed confirmed flush now publishes the fatal engine state under the WAL-writer lock, releases that lock, and only then performs synchronous teardown. A deterministic barrier test covers the prior checkpoint/index-lock versus writer-lock deadlock and verifies recovery.
3. Reader-directory enumeration now distinguishes absent from unreadable. Unreadable or malformed registries return an unknown state and skip all checkpoint work. A live external-reader test injects `UnauthorizedAccessException` and verifies its snapshot and WAL remain untouched.
4. Queries with `ReadTransform` bypass speculative buffering, and the reader no longer invokes the transform after EOF. A regression test verifies each callback runs exactly once.

I also corrected the backup wording: copying live data/WAL files independently is not presented as safe. The updated measurement includes the required legacy anchor (112 KiB vs 1,320 KiB, 91.5% less).

Local validation:
- Release solution build: success (netstandard2.0/net8/net10 plus test targets)
- net8 suite: 3077 passed, 0 failed, 7 existing skips
- focused new/reclamation coverage: green
- `python3 scripts/test-vector-compatibility.py`: green against 5.0.21, plain + encrypted
- changed-file size check and `git diff --check`: green

I will keep watching the full CI matrix and new review feedback.

File: tools/VectorCompatibility/Current/WalReuseCompatibility.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                Task.Run(() =>
+                {
+                    database.BeginTrans();
+                    victim.Update(Documents(99));
+                    database.Rollback();
+                }).GetAwaiter().GetResult();
+                var afterAbandoned = new FileInfo(LogName(source)).Length;
+                if (afterAbandoned != beforeAbandoned + 8192)
+                    throw new Exception("The abandoned transaction must reuse holes after its tail anchor: " +
+                        (afterAbandoned - beforeAbandoned));
```
[the remark this answers is not in the document]
Comment (JKamsker) [reply tools/VectorCompatibility/Current/WalReuseCompatibility.cs]: Fixed in eba7c7098. The fixture now allocates the delayed low-ID transaction first, writes and abandons a higher-ID transaction while the low transaction waits, verifies that the abandoned transaction grew the WAL by only its one tail page (so later frames used reclaimed slots), then resumes the low transaction. It asserts provisionalLow < abandonedHigh < finalLow and directly scans the plain/encrypted WAL to require both the abandoned ID and final physical tail/max ID == finalLow. LiteDB 5.0.21 then writes, checkpoints, and verifies the abandoned victim values remain unchanged. The production rebase prevents the collision rather than deliberately emitting a corrupt vector.

File: docs/mvcc-checkpoint.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+1. Flush the WAL, copy the safe page images, and durably flush the data file.
+2. Invalidate obsolete cached frames and overwrite their WAL slots with zero
+   pages, without moving any retained offset.
+3. Durably flush the cleared slots before publishing them to the free-slot pool.
+4. Append one unconfirmed frame for each new transaction before letting its later
+   frames reuse eligible slots under the WAL writer lock. This physical-tail
+   anchor preserves the maximum transaction ID for legacy recovery even if the
+   transaction is abandoned. Failed reused writes are not immediately returned
+   to the pool; reopening only recovers slots that are entirely zero. Abandoned
+   nonzero frames remain until full reset. Transaction-private safepoint reuse
```
[the remark this answers is not in the document]
Comment (JKamsker) [reply docs/mvcc-checkpoint.md]: Fixed in eba7c7098. Transaction IDs are now mutable transaction state. Under the WAL-writer lock, DiskService tracks the greatest physically observed ID; if a delayed lower-ID transaction arrives, it allocates a new ID above that maximum, durably appends a new-ID tail anchor, rewrites any earlier unconfirmed frames to the same ID, and only then permits confirmation. Snapshot validation and WAL-index confirmation both use the rebased ID. Plain/encrypted tests cover both a later safepoint and the confirmation-only path after earlier safepoints.

**Comment by JKamsker** [comment]
Follow-up review corrections are in `eba7c7098`.

- WAL transaction IDs now follow physical writer order, not allocation order. A delayed lower-ID transaction is rebased under the writer lock; its new-ID tail anchor is durably flushed before earlier unconfirmed frames are rewritten, and confirmation uses the rebased ID.
- Plain/encrypted unit coverage exercises rebasing during a later safepoint and the confirmation-only path after earlier safepoints.
- The 5.0.21 fixture now constructs the exact low-ID/high-ID arrival inversion, proves the abandoned higher-ID transaction used reclaimed slots, scans the plain/encrypted WAL to verify the rebased ID is both maximum and at the physical tail, then runs the legacy write/checkpoint/reopen validation.
- The failed-reused-write test now checks `cold` in both recovery modes, and the flush-deadlock test propagates checkpoint failures.
- All four CodeRabbit threads from review `5262158371` are addressed/resolved.

Local validation at this head:
- Release solution build: success, 0 errors
- net8 suite: 3,079 passed, 0 failed, 7 existing skips
- focused WAL/reclamation/durability tests: green, plain + encrypted
- `python3 scripts/test-vector-compatibility.py`: green through LiteDB 5.0.21, plain + encrypted
- changed-file size check and `git diff --check`: green

The preceding CI run had two isolated runner-bound failures after all other jobs passed: the full Windows latest x64/.NET 10 job hit its exact five-minute step timeout while tests were still progressing, and one Windows latest x86/.NET 8 child process hit its 30-second test timeout (the same cross-process matrix passed on the other 11 combinations). I am watching the fresh run before changing limits.

**Comment by JKamsker** [comment]
Final status on `e8a5da8cd`:

- Merged current `dev` (`faad53821`) and integrated its deleted-page-tail change with mutable/rebased WAL transaction IDs.
- Full GitHub Actions workflow is green: **47/47 workflow jobs passed**. The separate CodeRabbit check also passed, for **48 successful PR checks total**.
- The two combinations that timed out on the preceding run both passed: Windows latest x64/.NET 10 full suite and Windows latest x86/.NET 8 cross-process tests.
- CodeRabbit re-reviewed the final head, confirmed the transaction-ID collision concern is addressed, and left no unresolved review threads.
- Local merged-tree validation: Release solution build succeeds; net8 has **3,099 passed, 0 failed, 7 existing skips**; LiteDB 5.0.21 compatibility passes plain and encrypted.

**Comment by JKamsker** [comment]
Closing in favor of #3000: https://github.com/litedb-org/LiteDB/pull/3000, now part of native GitHub stack #3001.

The replacement starts from the same commit. This PR retains its review history; the original branch is unchanged and its head is also preserved at `codex/backup/storage-stack-20260922/pr-2936`. Please continue review and implementation on #3000.

#3000 remains draft pending the checksum-compatible reclamation protocol and integration tests described there.
